File: testbench/drivers/browser.mjs
/**
 * Generic headless-Chromium UI smoke (Playwright). Driven by a JSON config passed
 * as argv[2] (see drivers/browser.py). Loads the URL, clicks each declared view,
 * checks no console/page errors, runs optional click+assert interactions, and
 * prints a final JSON line: {checks:[{name,ok,detail}]}.
 *
 * Run from a dir with `node_modules/playwright`:
 *   node browser.mjs '{"url":"http://localhost:8000","views":[...]}'
 */
import { chromium } from 'playwright';

const cfg = JSON.parse(process.argv[2] || '{}');
const url = cfg.url || 'http://localhost:8000';
const checks = [];
const IGNORE = [/favicon/i];
let current = 'load';

const browser = await chromium.launch();
const page = await browser.newPage({ viewport: { width: 1400, height: 900 } });
const errors = [];
page.on('console', (m) => { if (m.type() === 'error' && !IGNORE.some((r) => r.test(m.text()))) errors.push({ view: current, text: m.text() }); });
page.on('pageerror', (e) => errors.push({ view: current, text: 'PAGEERROR: ' + e.message }));

try {
  const resp = await page.goto(url, { waitUntil: 'networkidle', timeout: 30000 });
  checks.push({ name: `load ${url}`, ok: !!(resp && resp.ok()), detail: 'HTTP ' + (resp ? resp.status() : '?') });
  await page.waitForTimeout(800);

  for (const v of cfg.views || []) {
    current = v.name;
    const before = errors.length;
    try {
      if (v.click) await page.click(v.click, { timeout: 5000 });
      await page.waitForTimeout(v.wait || 1000);
      const newErr = errors.length - before;
      checks.push({ name: `view '${v.name}' renders`, ok: newErr === 0, detail: newErr ? `${newErr} console error(s)` : 'clean' });
    } catch (e) {
      checks.push({ name: `view '${v.name}' renders`, ok: false, detail: e.message });
    }
  }

  for (const it of cfg.interactions || []) {
    current = it.name;
    try {
      if (it.goto_view) await page.click(it.goto_view, { timeout: 4000 }).catch(() => {});
      await page.waitForTimeout(500);
      if (it.click) await page.click(it.click, { timeout: 5000 });
      await page.waitForTimeout(it.wait || 1500);
      let ok = true, detail = 'done';
      if (it.expect_selector) {
        const txt = (await page.textContent(it.expect_selector).catch(() => '')) || '';
        ok = txt.trim().length > 0; detail = txt.slice(0, 60);
      }
      checks.push({ name: `interaction '${it.name}'`, ok, detail });
    } catch (e) {
      checks.push({ name: `interaction '${it.name}'`, ok: false, detail: e.message });
    }
  }
} finally {
  await browser.close();
}

const fails = checks.filter((c) => !c.ok).length;
console.error(`browser: ${checks.length - fails}/${checks.length} ok, ${errors.length} console error(s)`);
console.log(JSON.stringify({ checks }));
process.exit(fails ? 1 : 0);

File: testbench/drivers/browser.py
"""Browser / UI driver — headless-Chromium smoke of a web UI (generic).

Shells out to the Node script `browser.mjs` (Playwright). Verifies that every
declared view renders with no console/page errors, and runs optional click+assert
interactions. Gracefully skips if Node/Playwright isn't available.

Config (suites.browser):
  url: "http://localhost:8000"
  playwright_dir: "/tmp/pwtest"     # dir containing node_modules/playwright
                                    # (or set env TESTBENCH_PLAYWRIGHT_DIR)
  views:
    - {name: "engine",    click: ".tab-btn[data-tab='engine']"}
    - {name: "knowledge", click: ".tab-btn[data-tab='knowledge']"}
  interactions:
    - {name: "coach-preview", click: "button:has-text('Preview cue')", expect_selector: "#kp-result"}
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
from pathlib import Path

HERE = Path(__file__).resolve().parent


def run(cfg: dict) -> dict:
    checks = []

    def ck(name, ok, detail=""):
        checks.append(
            {"name": name, "status": "PASS" if ok else "FAIL", "detail": detail}
        )

    node = shutil.which("node")
    # env wins over config so a machine can override the (possibly placeholder) path.
    pw_dir = os.environ.get("TESTBENCH_PLAYWRIGHT_DIR") or cfg.get("playwright_dir")
    if not node:
        return {
            "passed": 0,
            "failed": 0,
            "skipped": 1,
            "checks": [
                {
                    "name": "browser",
                    "status": "PASS",
                    "detail": "node not found — skipped",
                }
            ],
            "note": "node/Playwright unavailable; browser suite skipped",
        }
    if not pw_dir or not (Path(pw_dir) / "node_modules" / "playwright").exists():
        return {
            "passed": 0,
            "failed": 0,
            "skipped": 1,
            "checks": [
                {
                    "name": "browser",
                    "status": "PASS",
                    "detail": "Playwright not installed (set playwright_dir / TESTBENCH_PLAYWRIGHT_DIR)",
                }
            ],
            "note": "Playwright unavailable; browser suite skipped",
        }

    payload = json.dumps(cfg)
    # ESM resolves `import 'playwright'` from the SCRIPT's directory (not cwd), so
    # copy the driver into pw_dir where node_modules/playwright lives.
    target = Path(pw_dir) / "_tb_browser.mjs"
    try:
        shutil.copy(HERE / "browser.mjs", target)
        proc = subprocess.run(
            [node, str(target), payload],
            cwd=pw_dir,
            capture_output=True,
            text=True,
            timeout=180,
        )
        out = proc.stdout.strip()
        # the script prints a final JSON line
        last = out.splitlines()[-1] if out else "{}"
        data = json.loads(last)
        for c in data.get("checks", []):
            ck(c["name"], c.get("ok", False), c.get("detail", ""))
        if not data.get("checks"):
            ck(
                "browser run",
                proc.returncode == 0,
                f"exit {proc.returncode}: {proc.stderr[-200:]}",
            )
    except Exception as e:  # noqa: BLE001
        ck("browser run", False, f"{type(e).__name__}: {e}")
    finally:
        try:
            target.unlink()
        except OSError:
            pass

    passed = sum(1 for c in checks if c["status"] == "PASS")
    failed = sum(1 for c in checks if c["status"] == "FAIL")
    return {"passed": passed, "failed": failed, "skipped": 0, "checks": checks}
